### DISCUSSION TITLE
Custom Plugin Response phase - how to set response body?

### DISCUSSION THREAD

**Discussion by sivasubramani300890** [Kong/kong#9509] [Help] [answered]
Kong 2.2.0 introduces new response phase (https://github.com/Kong/kong/blob/2.3.2/CHANGELOG.md#plugins-3) that according to the changelog allows to manipulate both the response headers and the response body in the same phase.

However, while kong.service.response.get_raw_body() and kong.service.response.get_body() are supported in this phase, there is no kong.response.set_raw_body() nor kong.response.set_body().
I am using python to develop custom plugin

How can the response body be manipulated in the response phase?

Thanks

**Comment by bungle** [comment, marked as answer]
There is (but that seems to only work on `body_filter`):
https://github.com/Kong/kong/blob/master/kong/pdk/response.lua#L605-L641

Though we need to check if that is available with external plugins.

**Reply by fffonion** [reply]
For external plugin:
- go: not yet, will be included in next release
- js/python: API is supported, but non-UTF8 decode-able body is not supported (like gzipped), will be fixed in next release

Do note this PDK added in kong 2.7.0, so apart from the external plugin server updated, you will also need Kong >= 2.7 to use those API.

**Comment by bungle** [comment]
Btw. the `response` phase is still ran on `access`, so you can use `kong.response.exit`.

**Reply by sivasubramani300890** [reply]
If I use kong.response.exit in the response phase, I am getting proper body response in Server. When I execute the API in postman it's not resulting the same.


**Getting Response Headers :**
Age →0
Connection →keep-alive
Content-Encoding →gzip
Content-Type →application/json
Date →Mon, 10 Oct 2022 11:02:12 GMT
ETag →"f76d5c3f-47dc-11ed-963d-819ea9e4b46d-2022-10-09T14:16:28.787000000Z--gzip"
Last-Modified →Sun, 09 Oct 2022 14:16:28 GMT
Transfer-Encoding →chunked
Vary →Accept, Accept-Encoding
Via →kong/2.8.1.3-enterprise-edition
X-Backend-History →facade_service
X-Cache →MISS
X-Cache-Origin →EAST
X-Kong-Proxy-Latency →5
X-Kong-Upstream-Latency →315
X-Shard-History →varnish_east2
X-Varnish →105063102
requestId →1234
x-hello-from-python →Python display Test Value
x-python-pid →16750

**Reply by sivasubramani300890** [reply]
When we use kong.response.exit in the access phase, we are getting the same responses in both the server and Postman. But we are unable to get_raw_body in the access phase.

for more details : 
**Actual Response :**
{
   "alldata":{
      "identifier":"BTU",
      "site":"au.bg",
      "role":"re"
   }
}

**Python Script :** 
def access(self, kong: kong.kong):
        host, err = kong.request.get_header("host")
        if err:
            pass  # error handling
        message = "hello"
        if 'message' in self.config:
            message = self.config['message']
        kong.response.set_header("x-hello-from-python", "Python display %s to %s" % (message, host))
        kong.response.set_header("x-python-pid", str(os.getpid()))

    def response(self, kong: kong):
        responseBody = kong.service.response.get_raw_body()
        convertedResponse = json.loads(responseBody[0])
        newDict = {"identifier":convertedResponse['devices'][0]['identifier'],"newSite":convertedResponse['devices'][0]['site'],"newkey":"newValue","updatedRole":convertedResponse['devices'][0]['role']}
        responseData = json.dumps({"Result":newDict})
        #kong.response.set_raw_body(responseData)
        #kong.log.err(f" { responseBody[0]} ")
        kong.response.exit(200, responseData)
        
**Response in Server :** 
{
   "Result":{
      "identifier":"BTU",
      "newSite":"au.bg",
      "newkey":"newValue",
      "updatedRole":"re"
   }
}
**Response in Postman :**
{
   "alldata":{
      "identifier":"BTU",
      "site":"au.bg",
      "role":"re"
   }
}


**Error log when executing API on postman :** 
2022/10/10 14:18:11 [error] 19240#0: *24001 [kong] mp_rpc.lua:315 [py-hello] no data while sending to client, client: 10.53.15.19, server: kong, request: "GET /ndsdeviceinfo?fqdn=abcdevice.net HTTP/1.1", host: "1.1.1.1:8000"
